### PR TITLE
All strings whose are valid array indicies always use the UINT32_IN_DESC format.

### PR DESCRIPTION
**Pull request by zherczeg**
Large performance gain (20% on nsieve-bits):

Benchmark | Perf (sec) |
----: | ----: | 
3d-cube.js | 0.935 -> 0.900 : +3.740% | 
3d-raytrace.js | 1.106 -> 1.078 : +2.514% | 
access-binary-trees.js | 0.568 -> 0.562 : +1.045% | 
access-fannkuch.js | 2.364 -> 2.221 : +6.053% | 
access-nbody.js | 1.077 -> 1.067 : +0.886% | 
bitops-3bit-bits-in-byte.js | 0.583 -> 0.584 : -0.306% | 
bitops-bits-in-byte.js | 0.868 -> 0.870 : -0.241% | 
bitops-bitwise-and.js | 1.061 -> 1.045 : +1.576% | 
bitops-nsieve-bits.js | 1.764 -> 1.410 : +20.046% | 
controlflow-recursive.js | 0.393 -> 0.394 : -0.349% | 
crypto-aes.js | 1.084 -> 0.992 : +8.545% | 
crypto-md5.js | 0.699 -> 0.674 : +3.556% | 
crypto-sha1.js | 0.677 -> 0.656 : +3.079% | 
date-format-tofte.js | 0.834 -> 0.790 : +5.335% | 
date-format-xparb.js | 0.422 -> 0.411 : +2.538% | 
math-cordic.js | 1.343 -> 1.316 : +2.054% | 
math-partial-sums.js | 0.743 -> 0.726 : +2.279% | 
math-spectral-norm.js | 0.591 -> 0.575 : +2.607% | 
string-base64.js | 1.755 -> 1.704 : +2.902% | 
string-fasta.js | 1.343 -> 1.245 : +7.342% | 
Geometric mean: | +3.870% | 

Binary size nearly unchanged (152644->152612)

### CODE REVIEW

File: jerry-core/ecma/base/ecma-helpers-string.c [hunk 1]
```diff
@@ -31,24 +31,18 @@
  * @{
  */
 
-/**
- * The ecma string ref counter should start after the container field.
- */
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-helpers-string.c]: Why did you remove these comments?

File: jerry-core/ecma/base/ecma-helpers-string.c [hunk 2]
```diff
@@ -90,6 +84,69 @@ ecma_init_ecma_string_from_magic_string_ex_id (ecma_string_t *string_p, /**< des
...
+ *         false otherwise
+ */
+static bool
+ecma_string_to_array_index (const lit_utf8_byte_t *string_p, /**< utf-8 string */
+                            lit_utf8_size_t string_size, /**< string size */
+                            uint32_t *result_p) /**< [out] converted value */
+{
+  JERRY_ASSERT (string_size > 0 && *string_p >= LIT_CHAR_0 && *string_p <= LIT_CHAR_9);
+
+  if (*string_p == LIT_CHAR_0)
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-string.c]: No, a[1] === a["1"] but a[1] !== a["001"]

File: jerry-core/ecma/base/ecma-helpers-string.c [hunk 3]
```diff
@@ -90,6 +84,69 @@ ecma_init_ecma_string_from_magic_string_ex_id (ecma_string_t *string_p, /**< des
...
+    if (*string_p > LIT_CHAR_9 || *string_p < LIT_CHAR_0)
+    {
+      return false;
+    }
+
+    index = (index * 10) + (uint32_t) (*string_p++ - LIT_CHAR_0);
+  }
+  while (string_p < string_end_p);
+
+  if (string_size < ECMA_MAX_CHARS_IN_STRINGIFIED_UINT32)
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-helpers-string.c]: I think at this point this is check is not necessary.

File: jerry-core/ecma/base/ecma-helpers-string.c [hunk 4]
```diff
@@ -90,6 +84,69 @@ ecma_init_ecma_string_from_magic_string_ex_id (ecma_string_t *string_p, /**< des
...
+  {
+    *result_p = index;
+    return true;
+  }
+
+  /* Overflow must be checked as well when size is
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-helpers-string.c]: There is an early return for overflow at line 105.

File: jerry-core/ecma/base/ecma-helpers-string.c (hunk 1 above)
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-string.c]: Because asserts have comment descriptions. These comments are just duplications (leftovers when assserts have no descriptions).

File: jerry-core/ecma/base/ecma-helpers-string.c (hunk 4 above)
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-helpers-string.c]: My bad. I misunderstood the comment.

File: jerry-core/ecma/base/ecma-helpers-string.c (hunk 3 above)
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-string.c]: This is a tricky code. The index computation cannot overflow for numbers whose length <= 9 characters. But for exactly 10 character long numbers we need to check overflow. This is why we have many checks, and we need them.

File: jerry-core/ecma/base/ecma-helpers-string.c [hunk 5]
```diff
@@ -479,7 +547,14 @@ ecma_concat_ecma_strings (ecma_string_t *string1_p, /**< first ecma-string */
     data_p = (lit_utf8_byte_t *) (long_string_desc_p + 1);
   }
 
-  string_desc_p->hash = lit_utf8_string_hash_combine (string1_p->hash, utf8_string2_p, utf8_string2_size);
+  lit_string_hash_t hash_start = string1_p->hash;
+
+  if (string1_is_uint32)
+  {
+    hash_start = lit_utf8_string_calc_hash (uint32_to_string_buffer1, utf8_string1_size);
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-helpers-string.c]: Couldn't be `hash_start = string1_p->u.uint32_number;` like in `ecma_init_ecma_string_from_uint32`?
